ktor-server/ktor-server-plugins/ktor-server-metrics/jvm/src/io/ktor/server/metrics/dropwizard/DropwizardMetrics.kt

```diff
@@ -10,133 +10,113 @@ import com.codahale.metrics.jvm.*
-                call.attributes.put(
-                    routingMetricsKey,
-                    RoutingMetrics(name, context)
-                )
-            }
+private data class CallMeasure constructor(val timer: Timer.Context)
+private val measureKey = AttributeKey<CallMeasure>("metrics")
 
-            pipeline.environment.monitor.subscribe(Routing.RoutingCallFinished) { call ->
-                val routingMetrics = call.attributes.take(routingMetricsKey)
-                val status = call.response.status()?.value ?: 0
-                val statusMeter = plugin.registry.meter(name(plugin.baseName, routingMetrics.name, status.toString()))
-                statusMeter.mark()
-                routingMetrics.context.stop()
-            }
+private object Monitoring : Hook<(ApplicationCall) -> Unit> {
+    override fun install(application: ApplicationCallPipeline, handler: (ApplicationCall) -> Unit) {
+        application.intercept(ApplicationCallPipeline.Monitoring) {
+            handler(call)
+        }
+    }
+}
 
-            return plugin
+private object AfterCall : Hook<(ApplicationCall) -> Unit> {
+    override fun install(application: ApplicationCallPipeline, handler: (ApplicationCall) -> Unit) {
+        application.intercept(ApplicationCallPipeline.Monitoring) {
+            try {
+                proceed()
+            } finally {
+                handler(call)
+            }
         }
     }
+}
 
-    private data class CallMeasure(val timer: Timer.Context)
+public val DropwizardMetrics: ApplicationPlugin<Application, DropwizardMetricsConfig, PluginInstance> =
+    createApplicationPlugin("DropwizardMetrics", ::DropwizardMetricsConfig) {
+        val duration = pluginConfig.registry.timer(name(pluginConfig.baseName, "duration"))
+        val active = pluginConfig.registry.counter(name(pluginConfig.baseName, "active"))
+        val exceptions = pluginConfig.registry.meter(name(pluginConfig.baseName, "exceptions"))
+        val httpStatus = ConcurrentHashMap<Int, Meter>()
+
+        if (pluginConfig.registerJvmMetricSets) {
+            listOf<Pair<String, () -> Metric>>(
+                "jvm.memory" to ::MemoryUsageGaugeSet,
+                "jvm.garbage" to ::GarbageCollectorMetricSet,
+                "jvm.threads" to ::ThreadStatesGaugeSet,
+                "jvm.files" to ::FileDescriptorRatioGauge,
+                "jvm.attributes" to ::JvmAttributeGaugeSet
+            ).filter { (name, _) ->
+                !pluginConfig.registry.names.any { existingName -> existingName.startsWith(name) }
+            }.forEach { (name, metric) -> pluginConfig.registry.register(name, metric()) }
+        }
 
-    private val measureKey = AttributeKey<CallMeasure>("metrics")
+        on(CallFailed) { _, _ ->
+            exceptions.mark()
+        }
 
-    private fun before(call: ApplicationCall) {
-        active.inc()
-        call.attributes.put(measureKey, CallMeasure(duration.time()))
-    }
+        on(CallRouted) { call ->
```

Comment:
I'm not sure why do we need these hooks when can use monitoring directly

Reply:
Just because these hoks encapsulate the boilerplate code and are much handier and consistend with new API